### PR TITLE
Allow for setting the Pickup Item of an Arrow

### PR DESCRIPTION
**Pull request by tl0x**
The feature proposed in issue [#7707](https://github.com/PaperMC/Paper/issues/7707).

### CODE REVIEW

**Comment by HexedHero** [comment]
Should this extend to other projectiles that can be picked up like Tridents and any future ones?

File: patches/api/0378-allow-setting-of-arrowpickup.patch [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+diff --git a/src/main/java/org/bukkit/entity/Arrow.java b/src/main/java/org/bukkit/entity/Arrow.java
+index 8814519df13c5e9475774f4bce4c8c6a462c2978..7f80a6d1a0f1f2d1889062a1f451e311bdc92e1a 100644
+--- a/src/main/java/org/bukkit/entity/Arrow.java
++++ b/src/main/java/org/bukkit/entity/Arrow.java
+@@ -2,6 +2,7 @@ package org.bukkit.entity;
+ 
+ import java.util.List;
+ import org.bukkit.Color;
++import org.bukkit.inventory.ItemStack;
```
Comment (Machine-Maker) [inline patches/api/0378-allow-setting-of-arrowpickup.patch:15]: Inline this import

File: patches/api/0378-allow-setting-of-arrowpickup.patch [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+@@ -94,4 +95,11 @@ public interface Arrow extends AbstractArrow {
+      * in a state with no Custom Effects and PotionType.UNCRAFTABLE
+      */
+     void clearCustomEffects();
++
++    /**
++     * Sets the ItemStack given when the arrow is picked up.
++     *
++     * @param itemStack
++     */
```
Comment (Machine-Maker) [inline patches/api/0378-allow-setting-of-arrowpickup.patch:28]: I would make some mention that this isn't persisted across entity unload/loads. aka its not saved.

File: patches/server/0891-allow-setting-of-arrowpickup.patch [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+index 53d0024daf6963ac4dab575666b0d6a74a39a958..00fd2849e4dd31b506645dbb9bd7128a453ac638 100644
+--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
++++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+@@ -603,6 +603,8 @@ public abstract class AbstractArrow extends Projectile {
+ 
+     public abstract ItemStack getPickupItem();
+ 
++    public abstract void setPickupItem(ItemStack stack);
```
Comment (Machine-Maker) [inline patches/server/0891-allow-setting-of-arrowpickup.patch:15]: I think the field should be here on AbstractArrow and then you can have a pickup item override for all things that extend AbstractArrow.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
You've got two server patches. You can do `git commit --amend` in `Paper-Server` to add changes to the last commit. 


Does this compile? I don't see the setPickupItem method implemented on ThrownTrident.

File: patches/api/0378-allow-setting-of-arrowpickup.patch (hunk 2 above)
Comment (u9g) [reply patches/api/0378-allow-setting-of-arrowpickup.patch:28]: why not save the itemstack as bytes in it’s pdc?

**Pull request closed without merging**